src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -602,6 +607,64 @@ private static Object bufferRef(Buffer buffer) {
         }
     }
 
+    // COPY_NATIVE_THRESHOLD must be a power of two and should be greater than 2^3
+    private static final long COPY_NATIVE_THRESHOLD = 1 << 6;
+
+    @ForceInline
+    public static void copy(AbstractMemorySegmentImpl src, long srcOffset,
+                            AbstractMemorySegmentImpl dst, long dstOffset,
+                            long size) {
+
+        Utils.checkNonNegativeIndex(size, "size");
+        // Implicit null check for src and dst
+        src.checkAccess(srcOffset, size, true);
+        dst.checkAccess(dstOffset, size, false);
+
+        if (size <= 0) {
+            // Do nothing
+        } else if (size < COPY_NATIVE_THRESHOLD && !src.overlaps(dst)) {
```

Comment:
This is basically a conservative test - correct? E.g. we might end up cosidering two segments as overlapping even when they are not (because of offsets) - but that's ok, I actually think it's a good pragmatic solution.

Reply:
Yes, it is conservative. We can tolerate overlaps in one direction (but not the other). We refrain from using these cases as they are unlikely, and checking would incur a performance price for all other (much more likely) cases.